lib/extras/packed_image.h

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#ifndef LIB_EXTRAS_PACKED_IMAGE_H_
+#define LIB_EXTRAS_PACKED_IMAGE_H_
+
+// Helper class for storing external (int or float, interleaved) images. This is
+// the common format used by other libraries and in the libjxl API.
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <string>
+#include <vector>
+
+#include "jxl/codestream_header.h"
+#include "jxl/types.h"
+#include "lib/jxl/base/status.h"
+#include "lib/jxl/common.h"
+
+namespace jxl {
+namespace extras {
+
+// Class representing an interleaved image with a bunch of channels.
+class PackedImage {
+ public:
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format)
+      : PackedImage(xsize, ysize, format, CalcStride(format, xsize)) {}
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format,
+              size_t stride)
+      : xsize(xsize),
+        ysize(ysize),
+        stride(stride),
+        format(format),
+        pixels(malloc(ysize * stride)),
+        pixels_size(ysize * stride) {}
+  // Construct the image using the passed pixel buffer. The buffer is owned by
+  // this object and released with free().
+  PackedImage(size_t xsize, size_t ysize, const JxlPixelFormat& format,
+              void* pixels, size_t pixels_size)
+      : xsize(xsize),
+        ysize(ysize),
+        stride(CalcStride(format, xsize)),
+        format(format),
+        pixels(pixels),
+        pixels_size(pixels_size) {
+    JXL_ASSERT(pixels_size >= stride * ysize);
+  }
+
+  PackedImage& operator=(PackedImage& other) = delete;
+  PackedImage& operator=(PackedImage&& other) {
+    xsize = other.xsize;
+    ysize = other.ysize;
+    flipped_y = other.flipped_y;
+    stride = other.stride;
+    format = other.format;
+    pixels_size = other.pixels_size;
+
+    if (pixels) free(pixels);
+    pixels = other.pixels;
+    other.pixels = nullptr;
+
+    return *this;
+  }
+
+  PackedImage(PackedImage& other) = delete;
+  PackedImage(PackedImage&& other)
+      : xsize(other.xsize),
+        ysize(other.ysize),
+        flipped_y(other.flipped_y),
+        stride(other.stride),
+        format(other.format),
+        pixels(other.pixels),
+        pixels_size(other.pixels_size) {
+    other.pixels = nullptr;
+  }
+
+  ~PackedImage() {
+    if (pixels) free(pixels);
+  }
+
+  // The image size in pixels.
+  size_t xsize;
+  size_t ysize;
+
+  // Whether the y coordinated is flipped (y=0 is the last row).
```

Comment:
coordinate, not coordinated

Also, do we need this? Couldn't we undo this in the pfm reader?

Reply:
We can probably get rid of it after updating the pfm reader/writer.

Reply:
I'm OK with doing it later.